* Regression

    * MAE
    * MSE
    * RMSE
    * R2 and Adj. R2
    

* Classification Metrics
    
    * Confusion Matrix
    
    * Accuracy Score
    
    * Precision
    
    * Recall
    
    * Specificity
    
    * Sensitivity
    
    * F1-Score
    
    * ROC / AUC
    
    
###### Null Accuracy - In case of an imbalanced dataset, if we just go by majority class and say that the majority class is correct then it's called Null accuracy.

In [14]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [15]:
df = pd.read_csv('dataset/diabetes.csv')
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
Pregnancies                 768 non-null int64
Glucose                     768 non-null int64
BloodPressure               768 non-null int64
SkinThickness               768 non-null int64
Insulin                     768 non-null int64
BMI                         768 non-null float64
DiabetesPedigreeFunction    768 non-null float64
Age                         768 non-null int64
Outcome                     768 non-null int64
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [17]:
df.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [18]:
features = ['Pregnancies','BMI','Insulin','Age']

X = df[features]
y = df['Outcome']

In [19]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.2,random_state=42)


In [20]:
from sklearn.linear_model import LogisticRegression

lr = LogisticRegression()

lr.fit(X_train,y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [21]:
y_pred = lr.predict(X_test)

In [22]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

print('Accruacy Score : ',accuracy_score(y_pred,y_test))

Accruacy Score :  0.6493506493506493


In [23]:
y_test.value_counts()

0    99
1    55
Name: Outcome, dtype: int64

In [24]:
99/(99+55)

0.6428571428571429

## Null Accuracy : 64.3%

## Logistic Regression : 64.9%

### Is Logistic regression a good model in this case? 

### No

In [12]:
print('Predicted : ',y_pred[:25])
print('Actual.   : ',y_test.values[:25])

Predicted :  [1 0 0 0 0 1 0 0 0 1 0 1 0 0 0 0 0 0 0 0 0 0 1 0 1]
Actual.   :  [0 0 0 0 0 0 0 0 0 0 1 0 1 0 0 1 0 0 1 1 0 0 0 0 0]


## Confusion Matrix

<img src='img/09_confusion_matrix_1.png'>




<img src='img/09_confusion_matrix_2.png'>




* TN - ppl who did not have diabetes and were predicted to not have diabetes (118)

* TP - ppl who had diabetes and were predicted to ave diabetes (15)

* FP - Incorrectly predicted that person did have diabetes (12)
    * Type 1 Error

* FN - Incorrectly predicted that person did not have diabetes (47)
    * Type 2 Error

In [13]:
133/192

0.6927083333333334

In [26]:
cm = confusion_matrix(y_pred,y_test)

TP = cm[1,1]
TN = cm[0,0]
FP = cm[0,1]
FN = cm[1,0]

print(f'TN: {TN} TP: {TP} FP : {FP} FN: {FN}')

TN: 80 TP: 20 FP : 35 FN: 19


#### Accuracy Score

In [29]:
print(f'Accuracy Score : {(TP+TN)/(TP+TN+FP+FN)}')
print('Accruacy Score : ',accuracy_score(y_pred,y_test))

Accuracy Score : 0.6493506493506493
Accruacy Score :  0.6493506493506493


#### Misclassification Rate

In [31]:
# 1-accuracy_score

misclassification_rate = 1 - accuracy_score(y_pred,y_test)
misclassification_rate

0.35064935064935066

### Precision = TP / (TP + FP)

In asteroid prediction problem, Precision = 0


Precision is used when we want to be very sure about our results.

### Recall = TP/(TP + FN)

Recall = 0 asteroid problem

Recall is used when we want to capture all positives.

### F1 Score = Harmonic mean of precision and recall

(2 * (precision * recall) / (precision + recall))

In [32]:
from sklearn.metrics import f1_score

y_actual = [0,1,1,0,1,1] 
y_pred =   [0,0,1,0,0,1]

f1_score(y_pred,y_actual)

0.6666666666666666

#### Specificity = TN / (TN+ FP) - How specific the classifier in predicting negative instances?


#### Senstivity = 

* Spam Filter 
    * Optimizing for precision or specificity
* Fraud Detection
    * Optimize for Senstivity or Recall

# Great Job!